torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,255 @@
+          aten::slice,
+          {linear_node->output(), neg1, slice_start_val, slice_end_val, one});
+      slice->insertAfter(linear_node);
+      orig_node->replaceAllUsesWith(slice);
+      orig_node->destroy();
+
+      slice_start = slice_end;
+      slice_start_val = slice_end_val;
+    }
+  }
+
+  bool isNonZeroDimEqual(Tensor& tensor_a, Tensor& tensor_b) {
+    if (tensor_a.dim() != tensor_b.dim()) {
+      return false;
+    }
+    for (int64_t i = 1; i < tensor_a.dim(); i++) {
+      if (tensor_a.size(i) != tensor_b.size(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Check the linear_layer_group of a tensor to find ones that can be
+  // combined
+  void collectAndMergeLinearLayers(std::list<Node*>& linear_layer_group) {
+    while (!linear_layer_group.empty()) {
+      auto it_first = linear_layer_group.begin();
+      Node* base_node = *it_first;
+
+      std::vector<Node*> compatible_layers;
+      compatible_layers.push_back(base_node);
+      // removing the node from future consideration, as we will only move
+      // nodes forward to keep code simple.
+      linear_layer_group.erase(it_first);
+
+      auto base_weight =
+          constant_as<Tensor>(base_node->namedInput("weight")).value();
+      auto base_bias =
+          constant_as<Tensor>(base_node->namedInput("bias")).value();
+
+      // Now iterate over the rest of the users of the set to
+      // see if there is anything that we can coaleasce `base_node` with.
+      for (auto it = linear_layer_group.begin();
+           it != linear_layer_group.end();) {
+        auto node = *it;
+        auto weight = constant_as<Tensor>(node->namedInput("weight")).value();
+        auto bias = constant_as<Tensor>(node->namedInput("bias")).value();
+
+        // For now we will just keep it simple and require matching types
+        // Type promotion mgiht cause performance to actually decrease.
+        if (base_weight.dtype() != weight.dtype() ||
+            base_weight.device() != weight.device() ||
+            base_bias.dtype() != bias.dtype() ||
+            base_bias.device() != bias.device()) {
+          it++;
+          continue;
+        }
+
+        if (!isNonZeroDimEqual(base_weight, weight) ||
+            !isNonZeroDimEqual(base_bias, bias)) {
+          it++;
+          continue;
+        }
+
+        bool can_move_before_all = true;
+        for (auto n : compatible_layers) {
+          can_move_before_all &=
+              getAliasDb()->couldMoveBeforeTopologically(node, n);
+        }
+        if (!can_move_before_all) {
+          it++;
+          continue;
+        }
+
+        // Found a node that is eligible for combination
+        auto prev = it;
+        it++;
+        linear_layer_group.erase(prev);
+        compatible_layers.push_back(node);
+      }
+      if (compatible_layers.size() == 1) {
+        continue; // No other layers to merge
+      }
+      mergeLinearLayers(base_node, compatible_layers);
```

Comment:
if base node is in `compatible_layers` it makes the api a little confusing that we're passing it in twice..